Blog Post: Adding New Distribution(s) in OpenSearch Project

peterzhuamazon opened the pull request:
### Description
Blog Post: Adding New Distribution(s) in OpenSearch Project
 
### Issues Resolved
https://github.com/opensearch-project/project-website/issues/968

### Check List
- [x] Commits are signed per the DCO using --signoff


By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

bbarani commented:
Don't merge this PR until https://github.com/opensearch-project/project-website/pull/1015 goes out since this blog has reference to that blog.

peterzhuamazon commented:
Hi @natebower @krisfreedain @bbarani Can you guys review this blog?
It has already been reviewed by @vagimeli before making into this PR.

Thanks!

natebower commented:
> Hi @natebower @krisfreedain @bbarani Can you guys review this blog? It has already been reviewed by @vagimeli before making into this PR.
> 
> Thanks!

Hi @peterzhuamazon @bbarani, I don't have an editorial request for this. Can you please submit one at https://form.asana.com/?k=C8mTcYB8-BAqQXZAy6yshQ&d=8442528107068%20form? Thanks.

peterzhuamazon commented:
> > Hi @natebower @krisfreedain @bbarani Can you guys review this blog? It has already been reviewed by @vagimeli before making into this PR.
> > Thanks!
> 
> Hi @peterzhuamazon @bbarani, I don't have an editorial request for this. Can you please submit one at https://form.asana.com/?k=C8mTcYB8-BAqQXZAy6yshQ&d=8442528107068%20form? Thanks.

Hi @natebower thanks for the comments and link.
I just discussed with Patti and she will submit the request on asana.
We are currently looking for a 10/27 release of this blog.

Thanks.

pajuric commented:
> > > Hi @natebower @krisfreedain @bbarani Can you guys review this blog? It has already been reviewed by @vagimeli before making into this PR.
> > > Thanks!
> > 
> > 
> > Hi @peterzhuamazon @bbarani, I don't have an editorial request for this. Can you please submit one at https://form.asana.com/?k=C8mTcYB8-BAqQXZAy6yshQ&d=8442528107068%20form? Thanks.
> 
> Hi @natebower thanks for the comments and link. I just discussed with @pajuric and she will submit the request on asana. We are currently looking for a 10/27 release of this blog.
> 
> Request has been opened in Asana for editorial review, as we have an opening in the schedule to publish.

natebower reviewed (commented):
@peterzhuamazon Please see my comments and changes, and let me know if you have any questions. Because a lot of my changes/comments are grammatical in nature, it would probably be best to have @vagimeli address them. Thanks!

natebower reviewed (commented):
@peterzhuamazon Please see my comments and changes, and let me know if you have any questions. Because a lot of my changes/comments are grammatical in nature, it would probably be best to have @vagimeli address them. Thanks!

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+With `TAR` as the initial distribution, we were able to wrap the archive into a container as part of the `Docker` release and then expand into multiple architectures including `x64` and `arm64`. Development of other distributions was delayed because the fundamental code base involved much hard coding to extend it.
+
+OpenSearch continuously improves. The community has asked for `RPM` packages, as it is an easy approach for LINUX users to deploy, run, and test OpenSearch and OpenSearch Dashboards clusters without complex manual setup. We see the demands and benefits of this new distribution and have prioritized its implementation. This effort led to a two-month-long project involving overhauling code and carry ...
```
@natebower I will change this to 
```
The community has asked for `RPM` package, as it is an easy way for LINUX users to deploy......
```

I will add the `the` as you suggested.

natebower replied:
It should either be "The community has asked for `RPM` packages, as they are an easy way..." or "The community has asked for an `RPM` package, as it is an easy way..."

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+## Getting started
+
+The [opensearch-build](https://github.com/opensearch-project/opensearch-build) repository (repo) is the centerpiece of artifacts generation for the OpenSearch project. Most automation and pipelines are running the code of this repo. To answer `where to start` the implementation of a new distribution (`RPM` for purposes in this blog), we need to understand the workflow in the opensearch-build repo ...
```
I will change `are running` to `run.

natebower replied:
> @natebower it should be plural as we have many different artifacts coming of the build script of this repo.

However, if we're referring to the general act of "artifact generation," it should be singular, even if we're talking about multiple artifacts. For example, we would say that someone is involved in "blog writing" and not "blogs writing."

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+The OpenSearch project’s final release artifacts (non-SNAPSHOT) have two formats: with plugins (normal format) and without plugins (min format, only for TAR). This naturally separates OpenSearch into two different parts: the OpenSearch engine and the related plugins. The engine is based on the [OpenSearch](https://github.com/opensearch-project/OpenSearch) main repository. The related plugins are  ...
+
+Three stages are involved in generating OpenSearch artifacts: *Build*, *Assemble*, and *Test*.
```
Will change * to `.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+The corresponding directories: `build_workflow`, `assemble_workflow`, and `test_workflow` contain the code for the `build`, `assemble`, `test` stages, respectively. In this blog, we focus on the `build` and `assemble` stages, as they are crucial to implementing a new OpenSearch distribution.
+
+While we are using Jenkins to execute the opensearch-build repo code for automated distribution builds, this blog doesn’t go into depth on that topic; see [OpenSearch automated build system is now live](https://opensearch.org/blog/community/2022/10/public-jenkins/) for more information. In this blog, we focus on running the `build` and `assemble` code on a local environment with the new distribut ...
```
Change to `on a local environment with necessary changes`.

peterzhuamazon replied:
Actually `on a local environment` is enough.
Since the entire post following is talking about what to do on a local env.

peterzhuamazon replied:
`in`

natebower replied:
`in a local environment with the necessary changes`

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+**Example: Simplified flow to generate artifacts on Jenkins**
+<img src="/assets/media/blog-images/2022-10-27-adding-new-distributions-in-opensearch-project/jenkins_pipeline_flowchart.png" alt="Jenkins Pipeliens Flowchart"/>{: .img-fluid }
+As shown in the figure above, after contributors commit their code or merge pull requests (PRs) in [opensearch-project](https://github.com/opensearch-project) repo, the pipelines on our [Public Jenkins](https://build.ci.opensearch.org/) will be triggered. From there, `build_workflow`, `assemble_workflow`, and `test_workflow` will be run sequentially and then the final products will be promoted to ...
```
`Cluster` is the missing word but we normally just call is `Public Jenkins` because `Jenkins` is already the name of the software.

natebower replied:
Ok, then no `the` beforehand. Just "the pipelines on Public Jenkins". Otherwise it would need to be "the pipelines on the Public Jenkins cluster".

peterzhuamazon replied:
Updated.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+## How to run build_workflow
+
+A common question asked during the `build_workflow` is “How do we run it?”. To answer that question, it’s necessary to understand what available parameters we have access to control the output.
```
confirmed same meaning.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+```
+The following parameters are available in `build.sh`.
+
+| name                    | description                                     |
````
They dont as they are part of the values directly taken from README on purpose.

peterzhuamazon replied:
On a second thought they look better when capitalized.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+
+```
+
+The figure above shows available parameters in `build_workflow`, before `RPM` is implemented. We do not have a `--distribution` parameter, as we were using `--platform` parameter to hardcode the distribution type. Initially, we assumed if `platform=linux, distribution=tar`. However, `RPM` also runs on LINUX hosts. 
````
Not really sure how to make it better.
What I want to describe is we used to assume `distribution` type based on `platform` type.
But that is not true anymore since we introduce RPM, because RPM can also run on linux platform.

Therefore, we need a new `--distribution` parameter to explicitly define the distribution type.

natebower replied:
The figure above shows the parameters available in `build_workflow` before `RPM` is implemented. There is no `--distribution` parameter, as the `--platform` parameter is used to hardcode the distribution type. Initially, it was assumed that if `platform=linux`, then `distribution=tar`.

peterzhuamazon replied:
a

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+```
+
+
+Next, we need to add a new parameter to handle different distribution.
````
`the different distributions`

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+| -d, --distribution      | Specify distribution to build, default is `tar`. |
+```
+
+When running `./build.sh`, internally it runs `./run.sh` to check dependencies on python/pip, which then runs `src/run_build.py`. Note, `src/run_build.py` is the entry point to `build_workflow` and starts the building process, including handling the user-defined parameters (see [run_build.py](https://github.com/opensearch-project/opensearch-build/blob/1.3.2/src/run_build.py)).
````
resolved.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+args = BuildArgs()
+```
+
+Next, we add the following code in [src/build_workflow/build_args.py](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-f5a380768626613f257a019bff13f8aec914e2b2983cea9d7a4552436eba4a45) using the `argparse` module.
````
Got it.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+Other files will need updates to support this new parameter. Examples include [src/build_workflow/build_target.py](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-22bbfc8727ab963ed470460b1996e12fbb9974a4caafb113fd2319ffa08a0a46) and [src/run_build.py](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-db6a1b8175460e03234674e4c8951029102a373968c ...
+
+When running build, we use an input manifest file (for this blog, it's `manifest/1.3.2/opensearch-1.3.2.yml`). This file consists of the components we want to build (the OpenSearch engine and related plugins). To learn more about the manifest files, read the code in [src/manifests](https://github.com/opensearch-project/opensearch-build/tree/1.3.2/src/manifests) for more information. Two other man ...
```
when running the `build_workflow`

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+Other files will need updates to support this new parameter. Examples include [src/build_workflow/build_target.py](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-22bbfc8727ab963ed470460b1996e12fbb9974a4caafb113fd2319ffa08a0a46) and [src/run_build.py](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-db6a1b8175460e03234674e4c8951029102a373968c ...
+
+When running build, we use an input manifest file (for this blog, it's `manifest/1.3.2/opensearch-1.3.2.yml`). This file consists of the components we want to build (the OpenSearch engine and related plugins). To learn more about the manifest files, read the code in [src/manifests](https://github.com/opensearch-project/opensearch-build/tree/1.3.2/src/manifests) for more information. Two other man ...
```
The result from build and assemble workflows. I will tweak it a bit better.

natebower replied:
I was just indicating that there should be a `the` before `results`.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+When running build, we use an input manifest file (for this blog, it's `manifest/1.3.2/opensearch-1.3.2.yml`). This file consists of the components we want to build (the OpenSearch engine and related plugins). To learn more about the manifest files, read the code in [src/manifests](https://github.com/opensearch-project/opensearch-build/tree/1.3.2/src/manifests) for more information. Two other man ...
+
+To add the distribution key-value pair in the build manifest so the `assemble_workflow` can use the build manifest to assemble the actual artifact in the given distribution format, we update [src/build_workflow/build_recorder.py](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-124059c5fa06b48851cc78f9260a851e37dec9dbbb3e9ea7cde945ad7b159c75) and [src/manifests/build_ma ...
```
Resolved.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+ self.data["build"]["distribution"] = target.distribution if target.distribution else "tar"
+```
+
+In `build_manifest.py`, `distribution` is `None` so that older `manifest.yml` without the `distribution` key-value pair is compatible with the new code:
````
addressed to:
```
In `build_manifest.py`, `distribution` is `None` so that the older `manifest.yml` files from previous releases without the `distribution` key-value pair are compatible with the new code:
```

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+ self.distribution: str = data.get('distribution', None)
+```
+
+After the above steps are complete, we are ready to craft the build scripts. We enter the thunderdome for [src/build_workflow/builder_from_source.py](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-e4f0b9f43bd4f3a5d5ca99e4ba9e7f38544f1078d2687d9ce5ae973df896e9ff). This file will call a subprocess that executes a shell script returned by the `ScriptFinder` function.  Th ...
````
I will remove but I was trying to describe the code being very complicated. 😄

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+)
+```
+
+The logic of `ScriptFinder.find_build_script` is in [src/paths/script_finder.py](https://github.com/opensearch-project/opensearch-build/blob/1.3.2/src/paths/script_finder.py#L57-L70) , which would allocate a build script named `build.sh` (same name but different script compares to the one in root directory of opensearch-build repository). This script can be found in one of the following paths:
````
revised:

(this script has the same name as the one in root directory of opensearch-build repository, but has very different content and is used for component specific build activity)

natebower replied:
(this script has the same name as the one in the root directory of the opensearch-build repository but contains very different content and is used for component-specific build activity)

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+The logic of `ScriptFinder.find_build_script` is in [src/paths/script_finder.py](https://github.com/opensearch-project/opensearch-build/blob/1.3.2/src/paths/script_finder.py#L57-L70) , which would allocate a build script named `build.sh` (same name but different script compares to the one in root directory of opensearch-build repository). This script can be found in one of the following paths:
+
+1. The opensearch-build repository `scripts/components/<Component Name>/build.sh`
+2. Component’s own repository root path `build.sh`
```
Resolved.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+1. The opensearch-build repository `scripts/components/<Component Name>/build.sh`
+2. Component’s own repository root path `build.sh`
+3. Component’s own repository `scripts/build.sh`
```
Resolved.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+For the [OpenSearch engine](https://github.com/opensearch-project/OpenSearch), because a [scripts/components/OpenSearch/build.sh](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-004d8135f0e1267ff669721dc64b44f84729dc6fe871c6c2ef02eb9d1e04959d) script exists in the opensearch-build repo, that script takes precedence and is returned by the `ScriptFinder` function. We now ...
+
+For [scripts/components/OpenSearch/build.sh](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-004d8135f0e1267ff669721dc64b44f84729dc6fe871c6c2ef02eb9d1e04959d), we add `distribution` as part of the variables and change how the corresponding Gradle task is called. In the OpenSearch engine, we use Gradle to compile the Java code into final artifact and component build.sh  ...
```
Revised, still awkward, will talk to Melissa on this later.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+## How to run assemble_workflow
+
+In `build_workflow`, you learned that OpenSearch artifacts are generated by calling the `root build.sh` and `component build.sh` as they interact with Gradle tasks (OpenSearch Dashboards is similar but interacts with Yarn). To combine or bundle the artifacts into a final product, we need to add more functions in `assemble_workflow`.
```
`scripts`.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+## How to run assemble_workflow
+
+In `build_workflow`, you learned that OpenSearch artifacts are generated by calling the `root build.sh` and `component build.sh` as they interact with Gradle tasks (OpenSearch Dashboards is similar but interacts with Yarn). To combine or bundle the artifacts into a final product, we need to add more functions in `assemble_workflow`.
```
Resolved.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+In `build_workflow`, you learned that OpenSearch artifacts are generated by calling the `root build.sh` and `component build.sh` as they interact with Gradle tasks (OpenSearch Dashboards is similar but interacts with Yarn). To combine or bundle the artifacts into a final product, we need to add more functions in `assemble_workflow`.
+
+You can use the command below to start the assemble process, as information in `build_workflow` is saved in build `manifest.yml` and we can pass the yml file as the input.
```
More accurate `YAML`.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+./assemble.sh rpm/builds/opensearch/manifest.yml
+```
+
+As in `build_workflow`, several files require support of the new `distribution` parameter,  such as [src/manifests/bundle_manifest.py](https://github.com/opensearch-project/opensearch-build/pull/1659/files#diff-8b97bcb19b01a2a0ed56bf0c59749b05bc42ca477be40ac20875bc8eaed0c16d).
````
Similar to `build_workflow`, several files changes are required to support the new `distribution` parameter in `assemble_workflow`

natebower replied:
"file changes"

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+self.distribution = build.distribution
+```
+
+Next, we need to understand the file structure. Unlike `build_workflow` where everything is build related, `assemble_workflow` is separated into bundle and dist. “Bundle” controls the general functions of assemble to create a bundle of the OpenSearch engine and related plugins. “Dist” contains specific business logic for types of distribution to repackage. For more information about the relations ...
````
different types of distributions

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+Next, we need to understand the file structure. Unlike `build_workflow` where everything is build related, `assemble_workflow` is separated into bundle and dist. “Bundle” controls the general functions of assemble to create a bundle of the OpenSearch engine and related plugins. “Dist” contains specific business logic for types of distribution to repackage. For more information about the relations ...
+
+Before we add new `RPM` implementation, we need to clean up existing code to support `distribution` of existing `tar` and `zip`. To do this, remove `from_path` function in [src/assemble_workflow/dist.py](https://github.com/opensearch-project/opensearch-build/pull/1659/files#diff-b6de2321e4e1b16c1d5c6a3c6c4ef51bb2a5a3473f116500a9f859e18f4c7d0a), as it hardcodes the distribution type based on the a ...
```
Revised see other comment on changes.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+Before we add new `RPM` implementation, we need to clean up existing code to support `distribution` of existing `tar` and `zip`. To do this, remove `from_path` function in [src/assemble_workflow/dist.py](https://github.com/opensearch-project/opensearch-build/pull/1659/files#diff-b6de2321e4e1b16c1d5c6a3c6c4ef51bb2a5a3473f116500a9f859e18f4c7d0a), as it hardcodes the distribution type based on the a ...
+
+```diff
+We need to remove all of these old logic
````
Before you implement the `RPM` distribution, you need to clean up the existing code to support older distributions such as `Tar` and `Zip`. To do this, remove `from_path` function in [src/assemble_workflow/dist.py](https://github.com/opensearch-project/opensearch-build/pull/1659/files#diff-b6de2321e4e1b16c1d5c6a3c6c4ef51bb2a5a3473f116500a9f859e18f4c7d0a), as it hardcodes the distribution type based on file extension.

natebower replied:
"the `from_path` function". "the file extension."

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+            tar.add(self.archive_path, arcname=os.path.basename(self.archive_path))
+```
+
+The `class DistTar` extends `class Dist` with two `@abstractmethod` to be implemented: extract and build. Extract takes the OpenSearch engine Tarball artifact from the build stage, extracts the content so we can install plugins, and adjusts the configuration. Build (not to be confused with `build_workflow`) wraps the extracted content after the extract process and regenerates a new tarball as the ...
````
```
The `DistTar` class extends `Dist` class
```

Other two addressed.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
```diff
@@ -0,0 +1,411 @@
+
+The `class DistTar` extends `class Dist` with two `@abstractmethod` to be implemented: extract and build. Extract takes the OpenSearch engine Tarball artifact from the build stage, extracts the content so we can install plugins, and adjusts the configuration. Build (not to be confused with `build_workflow`) wraps the extracted content after the extract process and regenerates a new tarball as the ...
+
+After extracting the content, additional refinements can be made in [src/run_assemble.py](https://github.com/peterzhuamazon/opensearch-build/blob/b0ac712e1ce6d10e901ab549d7335ee3b20c81b9/src/run_assemble.py#L42-L45) that was called by `assemble.sh` in the opensearch-build repo root directory.
```
addressed.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+    logging.info(f"Installed plugins: {bundle.installed_plugins}")
+```
+
+The flow runs as `assemble.sh` calls `run_assemble.py`, which calls `Bundles.create`. It then returns the class `bundle_opensearch.py`, which creates an instance of `BundleOpenSearch` that extends `Bundle`, and then `Bundle` internally calls `Dists.create`, which returns `DistTar`, and calls [extract](https://github.com/peterzhuamazon/opensearch-build/blob/b0ac712e1ce6d10e901ab549d7335ee3b20c81b9 ...
````
addressed.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+min_dist.extract(self.tmp_dir.name)
+```
+
+After extraction, we return to `run_assemble.py` to continue installation.
````
addressed.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+bundle.install_plugins()
+```
+
+If no plugin is installed, OpenSearch or OpenSearch Dashboards are referred to as `min`, meaning they are the `minimal` product. The two calls, `bundle.install_min()` and `bundle.install_plugins()`, act similarly to the `build()` call in [builder_from_source](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-e4f0b9f43bd4f3a5d5ca99e4ba9e7f38544f1078d2687d9ce5ae973df896e9f ...
````
build() is calling build.sh
install() is calling install.sh.

I will re-format this and talk to Melissa to make it less confusing.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+bundle.install_plugins()
+```
+
+If no plugin is installed, OpenSearch or OpenSearch Dashboards are referred to as `min`, meaning they are the `minimal` product. The two calls, `bundle.install_min()` and `bundle.install_plugins()`, act similarly to the `build()` call in [builder_from_source](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-e4f0b9f43bd4f3a5d5ca99e4ba9e7f38544f1078d2687d9ce5ae973df896e9f ...
````
I didnt see any diffs in this one tho.

natebower replied:
For larger blocks of text, it appears that GitHub does not display the diffs for some reason.

peterzhuamazon replied:
Apparently another `We/You` 😄

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+    super().install_plugin(plugin)
+```
+
+We now have completed installation, so we return to `run_assemble.py` to call the following remaining functions:
````
addressed.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+        BundleRpm(self.filename, self.path, self.min_path).build(name, dest, self.archive_path)
+```
+
+Similar changes will be added to [src/assemble_workflow/dists.py](https://github.com/opensearch-project/opensearch-build/pull/1726/files#diff-9d09f47256a1ba3eb99c43c336833f2854a076cef3e1ed048ab2a536aa0caea3). We then will implement `RPM` specific bundle code in [src/assemble_workflow/bundle_rpm.py](https://github.com/opensearch-project/opensearch-build/pull/1726/files#diff-51ea9acc525537d3b2ac2f4 ...
````
Addressed and make the last sentence readable.

peterzhuamazon commented on _posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md:
````diff
@@ -0,0 +1,411 @@
+```
+
+
+Congratulations! You have implemented the `RPM` distribution creation process within `opensearch-build` code base. 
````
addressed.

peterzhuamazon commented:
Thanks @natebower for the lengthy review.

I have addressed the issues and will review again with @vegimeli before I rebase the commits.

Still looking for a 10/27 publishing. Thanks.

vagimeli reviewed (approved):
@peterzhuamazon @natebower I reviewed comments and the resolutions to those comments. Looks good to me.

krisfreedain reviewed (approved):
thanks @peterzhuamazon !

peterzhuamazon commented:
> thanks @peterzhuamazon !

Thanks everyone!
